### ISSUE TITLE
Nested header on input

### ISSUE THREAD

**Issue by victoraugustolls** [danielgtaylor/huma#301]
I see that it is a common pattern with `Huma` that the body, both in the input and output, is to be created inside the nested struct field `Body`. Is it possible to do the same with headers and path/query parameters? Something like:

```go
type MyInput struct {
    Headers struct {
        UserIP string `header:"user-ip"`
    }
    Body struct {
        Name string `json:"name"`
    }
}
```

**Comment by danielgtaylor** [comment]
@victoraugustolls this isn't currently possible. They are just added at the top level, like:

```go
type MyInput struct {
    UserIP  string `header:"User-Ip"`
    Header2 string `header:"Header2"`
    Header3 int    `header:"Header3"`
    Body struct {
        Name string `json:"name"`
    }
}
```

Any reason you need additional grouping?

**Comment by victoraugustolls** [comment]
I was thinking about header reuse, which allows structures to be reused on inputs. Do you think it makes sense?

**Comment by bclements** [comment]
I really like the grouping idea myself. It follows your established pattern to use Body. If we are upvoting..I'd love this as a feature.

**Comment by Insei** [comment]
@victoraugustolls you can use composition in you case, like:
```go
type MyInputHeaders struct {
    UserIP  string `header:"User-Ip"`
    Header2 string `header:"Header2"`
    Header3 int    `header:"Header3"`
}
type MyInput struct {
   MyInputHeaders
   Body struct {
        Name string `json:"name"`
   }
}
```
It's a not perfect solution, but I already use this method for each paginated lists in my APIs.

**Comment by victoraugustolls** [comment]
I now we can it just not feels that follows a pattern as with the body like @bclements mentioned!

One example of how can this help is when we are accessing the values of the input. Imagine a case where I have everything: headers, body, path/query parameters...

In such a case I know I am using a body variable because I access it with `input.Body.Value`, but this is not true for the rest! In this scenario the user need to look at the input in the case that it does not know by heart.

**Comment by victoraugustolls** [comment]
If @danielgtaylor agrees with it, I can open a PR implementing this for other structures other than the body!

**Comment by danielgtaylor** [comment]
I'm not opposed to the idea; seems several people are interested in it as a code organization strategy. It'll likely be slightly less efficient as reflection is used to set the fields and you have to go into the `Headers` struct but it's a tiny difference. As long as it is optional this seems like it would be a nice feature to have.

**Comment by ssoroka** [comment]
I'd actually like to make Body optional, and I think it might make sense to make both Body and Headers optional.
I don't think it'd be too confusing to intuit what each parameter does from the tags, eg all `header` are headers, `json` are clearly body, (`path` and `query` things, right?), and anything left over can probably be assumed to be body.

**Comment by victoraugustolls** [comment]
@danielgtaylor given @ssoroka comment, I also feel like that for every possible part of the request we could have it in the root of the struct or in special fields like Body, Header, Query and Param. Do we need a discussion for this or would you be okay with this proposal?

**Comment by danielgtaylor** [comment]
@victoraugustolls @ssoroka I'm not sure this will work for the body. In a scenario like this the JSON unmarshal call would not know what to do:

```go
type ExampleRequest struct {
  Part1 struct {
    Field1 string `json:"field1"`
  }
  Part2 struct {
    Field2 string `json:"field2"`
  }
}
```

There's also the complication that fields without a `json` tag are still unmarshaled, and we need to support other formats like CBOR/YAML/etc.

**Comment by victoraugustolls** [comment]
@danielgtaylor fair enough, did not think it through before tagging you! Thanks for the quick response, will work on the previous approach. I had some rough recent couple of weeks at works but I think I can take the time now :)
